scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingMethodImpl.java

```diff
@@ -196,115 +197,128 @@ public int schedule() {
+                        return numberOfTaskStarted;
+                    }
 
-                logger.debug("required number of nodes : " + neededResourcesNumber);
-                if (neededResourcesNumber == 0 || tasksToSchedule.isEmpty()) {
-                    break;
-                }
+                    //get the next compatible tasks from the whole returned policy tasks
+                    LinkedList<EligibleTaskDescriptor> tasksToSchedule = new LinkedList<>();
+                    int neededResourcesNumber = 0;
 
-                NodeSet nodeSet = getRMNodes(jobMap, neededResourcesNumber, tasksToSchedule, freeResources);
+                    while (taskRetrievedFromPolicy.size() > 0 && neededResourcesNumber == 0) {
+                        //the loop will search for next compatible task until it find something
+                        neededResourcesNumber = getNextcompatibleTasks(jobMap,
+                                                                       taskRetrievedFromPolicy,
+                                                                       freeResources.size(),
+                                                                       tasksToSchedule);
+                    }
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("tasksToSchedule : " + tasksToSchedule);
+                    }
 
-                if (nodeSet != null) {
-                    freeResources.removeAll(nodeSet.getAllNodesUrls());
-                }
+                    logger.debug("required number of nodes : " + neededResourcesNumber);
+                    if (neededResourcesNumber == 0 || tasksToSchedule.isEmpty()) {
+                        break;
+                    }
 
-                //start selected tasks
-                Node node = null;
-                InternalJob currentJob = null;
-                try {
-                    while (nodeSet != null && !nodeSet.isEmpty()) {
-                        EligibleTaskDescriptor taskDescriptor = tasksToSchedule.removeFirst();
-                        currentJob = ((JobDescriptorImpl) jobMap.get(taskDescriptor.getJobId())).getInternal();
-                        InternalTask internalTask = currentJob.getIHMTasks().get(taskDescriptor.getTaskId());
+                    NodeSet nodeSet = getRMNodes(jobMap, neededResourcesNumber, tasksToSchedule, freeResources);
+
+                    if (nodeSet != null) {
+                        freeResources.removeAll(nodeSet.getAllNodesUrls());
+                    }
 
-                        if (currentPolicy.isTaskExecutable(nodeSet, taskDescriptor)) {
-                            //create launcher and try to start the task
-                            node = nodeSet.get(0);
+                    //start selected tasks
+                    Node node = null;
+                    InternalJob currentJob = null;
+                    try {
+                        while (nodeSet != null && !nodeSet.isEmpty()) {
```

Comment:
I would extract this whole while loop in a separate method, if possible, like "launchTasks" for readability

Reply:
Make sense I agree. I just tried to not reorganize existing code but I will extract it in a dedicated method if you agree.